[config] Refactored module hack to a normal class/instance pattern. Fixed docs

marscher opened the pull request:
Now each property of the Config class has a documentation. Moved the examples
    to doc/source/Configuration.rst (because numpydoc does not allow for extra
    sections).
    Removed all dirty hacks of a class instance pretending to be a module.
    
    Fixes #1089

marscher committed:
[config] Refactored module hack to a normal class/instance pattern. Fixed docs

Now each property of the Config class has a documentation. Moved the examples
to doc/source/Configuration.rst (because numpydoc does not allow for extra
sections).
Removed all dirty hacks of a class instance pretending to be a module.

Fixes #1089

marscher committed:
fixes

marscher committed:
[doc] amending changelog [ci skip]

franknoe commented:
Thanks. Also, it would be nice if we can set the loggin level via the 
class. Currently:

```python
import logging
root_logger = logging.getLogger('pyemma')
root_logger.level = logging.WARN
```

is a bit weird, because you have to know that there is a registered 
logger called pyemma - no way to know that without reading the config 
docs. Could we do something like this:

```python
import logging
from pyemma import config
config.logger.level = logging.WARN
```

or maybe there's an even less complex way of doing this, e.g.

```python
from pyemma import config
config.loglevel = "WARN"
```

The current need for this is that with cross-validation you might repeat 
clustering many times which creates a ton of these not-converged INFOs. 
I guess we shouldn't lower the logging level of these messages to be 
generally invisible, but one needs a simple and API-obvious way to turn 
them off when the procedure is repeated many times. Another option would 
be pass another argument to the estimator, such as logging=False.

Am 19/05/17 um 19:06 schrieb Martin K. Scherer:
> |Now each property of the Config class has a documentation. Moved the 
> examples to doc/source/Configuration.rst (because numpydoc does not 
> allow for extra sections). Removed all dirty hacks of a class instance 
> pretending to be a module. Fixes #1089 |
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/markovmodel/PyEMMA/pull/1095
>
>
>         Commit Summary
>
>   * [config] Refactored module hack to a normal class/instance
>     pattern. Fixed docs
>   * fixes
>
>
>         File Changes
>
>   * *M* doc/source/Configuration.rst
>     <https://github.com/markovmodel/PyEMMA/pull/1095/files#diff-0> (108)
>   * *M* doc/source/legal.rst
>     <https://github.com/markovmodel/PyEMMA/pull/1095/files#diff-1> (4)
>   * *M* pyemma/coordinates/tests/test_traj_info_cache.py
>     <https://github.com/markovmodel/PyEMMA/pull/1095/files#diff-2> (3)
>   * *M* pyemma/util/__init__.py
>     <https://github.com/markovmodel/PyEMMA/pull/1095/files#diff-3> (5)
>   * *R* pyemma/util/_config.py
>     <https://github.com/markovmodel/PyEMMA/pull/1095/files#diff-4> (182)
>
>
>         Patch Links:
>
>   * https://github.com/markovmodel/PyEMMA/pull/1095.patch
>   * https://github.com/markovmodel/PyEMMA/pull/1095.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQnQOOvhRodc2RHiX3X4uISYWG5ksks5r7cwbgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
Having the root logger in the config might be very confusing, because this interface only contains simple key, value stores.
For your specific problem of these non-converged warnings (which are in general very useful for the user), you should lower the log-level of your clustering instance. This can be easily done, because every estimator holds its own logger. So you can just write:

clustering_obj.logger.level = logging.WARN

Unfortunately the logging module does not support assigning strings to the level (has to be int).

marscher commented:
https://stackoverflow.com/questions/35679012/how-to-convert-python-logging-level-name-to-integer-code#35689599

marscher commented:
We could add the information on how to change log levels to the logger attribute. What do you think?

franknoe commented:
Hm, that would work, but then we can't use the API function. Instead of

```python
dtrajs = coordinates.cluster_kmeans(Y, k=nc).dtrajs
```

I would do

```python
kmeans = coordinates.KmeansClustering(k=nc)
kmeans.logger.setLevel('WARN')
kmeans.fit(Y)
```

then I would rather prefer to be able to set the logger level globally 
(globally just means notebook-wide, so it's a rather "local" notion of 
"global").


Am 19/05/17 um 19:34 schrieb Martin K. Scherer:
>
> https://stackoverflow.com/questions/35679012/how-to-convert-python-logging-level-name-to-integer-code#35689599
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302764710>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQswk464Ci5FU4IzHn76qEcs5PDuIks5r7dKQgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
That's true. Haven' thought of that case. But since Pythons logging system is hierachical one could do it globally like this:

import logging
coordinates_clustering_logger = logging.getLogger('pyemma.coordinates.clustering')
coordinates_clustering_logger.setLevel('WARN')

prior invoking any clustering estimators. So this will change the log level only for this particular subpackage.

I think what you want is a global "shut-up" option?


On 05/19/2017 12:38 PM, Frank Noe wrote:
> Hm, that would work, but then we can't use the API function. Instead of
> 
> ```python
> dtrajs = coordinates.cluster_kmeans(Y, k=nc).dtrajs
> ```
> 
> I would do
> 
> ```python
> kmeans = coordinates.KmeansClustering(k=nc)
> kmeans.logger.setLevel('WARN')
> kmeans.fit(Y)
> ```
> 
> then I would rather prefer to be able to set the logger level globally
> (globally just means notebook-wide, so it's a rather "local" notion of
> "global").
> 
> 
> Am 19/05/17 um 19:34 schrieb Martin K. Scherer:
>>
>> https://stackoverflow.com/questions/35679012/how-to-convert-python-logging-level-name-to-integer-code#35689599
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302764710>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGMeQswk464Ci5FU4IzHn76qEcs5PDuIks5r7dKQgaJpZM4Ngx21>.
>>
> 
> 
> -- 
> 
> ----------------------------------------------
> Prof. Dr. Frank Noe
> Head of Computational Molecular Biology group
> Freie Universitaet Berlin
> 
> Phone: (+49) (0)30 838 75354
> Web: research.franknoe.de
> Mail: Arnimallee 6, 14195 Berlin, Germany
> ----------------------------------------------
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302765967>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKZL9B8cAvOGli_qHVkNetM7MhF8xkCks5r7dOkgaJpZM4Ngx21>.
>

franknoe commented:
That's right. I think it's good to have the option to control individual 
loggers for people that know Python well, but for the average user who 
wants to do some sort of adaptive sampling, cross-validation or anything 
that involves repeated calls to functions that are verbose, it would be 
good to have a simple global "shut-up" switch.

Currently I am doing this:
```python
# switch off progress bars
from pyemma import config
config.show_progress_bars = False
# mute logger
import logging
root_logger = logging.getLogger('pyemma')
root_logger.level = logging.WARN
```

Maybe we can make a shortcut to do that? Something like:

pyemma.config.mute()
pyemma.config.unmute()

or even

pyemma.mute()
pyemma.unmute()

Maybe there's a better name for this than mute.

Am 19/05/17 um 19:43 schrieb Martin K. Scherer:
> That's true. Haven' thought of that case. But since Pythons logging 
> system is hierachical one could do it globally like this:
>
> import logging
> coordinates_clustering_logger = 
> logging.getLogger('pyemma.coordinates.clustering')
> coordinates_clustering_logger.setLevel('WARN')
>
> prior invoking any clustering estimators. So this will change the log 
> level only for this particular subpackage.
>
> I think what you want is a global "shut-up" option?
>
>
> On 05/19/2017 12:38 PM, Frank Noe wrote:
> > Hm, that would work, but then we can't use the API function. Instead of
> >
> > ```python
> > dtrajs = coordinates.cluster_kmeans(Y, k=nc).dtrajs
> > ```
> >
> > I would do
> >
> > ```python
> > kmeans = coordinates.KmeansClustering(k=nc)
> > kmeans.logger.setLevel('WARN')
> > kmeans.fit(Y)
> > ```
> >
> > then I would rather prefer to be able to set the logger level globally
> > (globally just means notebook-wide, so it's a rather "local" notion of
> > "global").
> >
> >
> > Am 19/05/17 um 19:34 schrieb Martin K. Scherer:
> >>
> >> 
> https://stackoverflow.com/questions/35679012/how-to-convert-python-logging-level-name-to-integer-code#35689599
> >>
> >> —
> >> You are receiving this because you commented.
> >> Reply to this email directly, view it on GitHub
> >> 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302764710>,
> >> or mute the thread
> >> 
> <https://github.com/notifications/unsubscribe-auth/AGMeQswk464Ci5FU4IzHn76qEcs5PDuIks5r7dKQgaJpZM4Ngx21>.
> >>
> >
> >
> > --
> >
> > ----------------------------------------------
> > Prof. Dr. Frank Noe
> > Head of Computational Molecular Biology group
> > Freie Universitaet Berlin
> >
> > Phone: (+49) (0)30 838 75354
> > Web: research.franknoe.de
> > Mail: Arnimallee 6, 14195 Berlin, Germany
> > ----------------------------------------------
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302765967>, 
> or mute the thread
> > 
> <https://github.com/notifications/unsubscribe-auth/AAKZL9B8cAvOGli_qHVkNetM7MhF8xkCks5r7dOkgaJpZM4Ngx21>.
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302767085>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQuYJfMSlGPAD8Za8xd6LQ5xHEmSpks5r7dTYgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
> Maybe we can make a shortcut to do that? Something like:
> 
> pyemma.config.mute()
> pyemma.config.unmute()
> 
> or even
> 
> pyemma.mute()
> pyemma.unmute()

That's a good idea. I'm going to add the mute property to the config, which can be True or False.
Are you really sure you want to have this in the global (pyemma) namespace?

franknoe commented:
It should be easily accessible. I don't mind if it's under config, but 
then the config should be automatically imported.

Am 19/05/17 um 20:35 schrieb Martin K. Scherer:
> > Maybe we can make a shortcut to do that? Something like:
> >
> > pyemma.config.mute()
> > pyemma.config.unmute()
> >
> > or even
> >
> > pyemma.mute()
> > pyemma.unmute()
>
> That's a good idea. I'm going to add the mute property to the config, 
> which can be True or False.
> Are you really sure you want to have this in the global (pyemma) 
> namespace?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302779145>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQhr08MTDTK4gJ3yCykQHSfCyxoc9ks5r7eDxgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
On 05/19/2017 01:41 PM, Frank Noe wrote:
> It should be easily accessible. I don't mind if it's under config, but
> then the config should be automatically imported.

pyemma.config is already there, so it is easily accessible.

franknoe commented:
P.S. if you use a mute property that could lead to conflicts. E.g. 
setting mute=True, but show_progress_bars=True. If it's a function that 
sets config values accordingly, it might be clearer.

Am 19/05/17 um 20:35 schrieb Martin K. Scherer:
> > Maybe we can make a shortcut to do that? Something like:
> >
> > pyemma.config.mute()
> > pyemma.config.unmute()
> >
> > or even
> >
> > pyemma.mute()
> > pyemma.unmute()
>
> That's a good idea. I'm going to add the mute property to the config, 
> which can be True or False.
> Are you really sure you want to have this in the global (pyemma) 
> namespace?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302779145>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQhr08MTDTK4gJ3yCykQHSfCyxoc9ks5r7eDxgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

franknoe commented:
So

import pyemma
pyemma.config.[property] = [value]

works without additional importing of config? That would be new I think.

Am 19/05/17 um 20:43 schrieb Martin K. Scherer:
> On 05/19/2017 01:41 PM, Frank Noe wrote:
> > It should be easily accessible. I don't mind if it's under config, but
> > then the config should be automatically imported.
>
> pyemma.config is already there, so it is easily accessible.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302780854>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQh4I9g509d5VjkDXur32dqa99O5qks5r7eK3gaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
On 05/19/2017 01:45 PM, Frank Noe wrote:
> So
> 
> import pyemma
> pyemma.config.[property] = [value]
> 
> works without additional importing of config? That would be new I think.
We have this already for a long time now.

Conflicts can be resolved easily, because the setter of mute also invokes the setter of show_progress_bars.

franknoe commented:
Am 19/05/17 um 20:50 schrieb Martin K. Scherer:
> On 05/19/2017 01:45 PM, Frank Noe wrote:
> > So
> >
> > import pyemma
> > pyemma.config.[property] = [value]
> >
> > works without additional importing of config? That would be new I think.
> We have this already for a long time now.
you're right, sorry.
> Conflicts can be resolved easily, because the setter of mute also 
> invokes the setter of show_progress_bars.
OK, that works.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1095#issuecomment-302782472>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQnG8x6U-u-YJt9WqO4yceoB_bInpks5r7eRcgaJpZM4Ngx21>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher committed:
[config] added mute option which silences progressbars and logging

Note that critical errors are still being logged, because we need a way to
inform users about serious errors.

marscher commented:
added and tested mute option.

marscher committed:
[doc] amending changelog [ci skip]
